pkg/genericapiserver/openapi/openapi.go

```diff
@@ -16,213 +16,138 @@ limitations under the License.
-	OpenAPIServePath = "/swagger.json"
-	OpenAPIVersion   = "2.0"
+	OpenAPIVersion = "2.0"
 )
 
 // Config is set of configuration for openAPI spec generation.
 type Config struct {
-	// SwaggerConfig is set of configuration for go-restful swagger spec generation. Currently
-	// openAPI implementation depends on go-restful to generate models.
-	SwaggerConfig *swagger.Config
+	// Path to the spec file. by convention, it should name [.*/]*/swagger.json
+	OpenAPIServePath string
+	// List of web services for this API spec
+	WebServices []*restful.WebService
+
+	// List of supported protocols such as https, http, etc.
+	ProtocolList []string
+
 	// Info is general information about the API.
 	Info *spec.Info
 	// DefaultResponse will be used if an operation does not have any responses listed. It
-	// will show up as ... "responses" : {"default" : $DefaultResponse} in swagger spec.
+	// will show up as ... "responses" : {"default" : $DefaultResponse} in the spec.
 	DefaultResponse *spec.Response
 	// List of webservice's path prefixes to ignore
 	IgnorePrefixes []string
 }
 
+// +k8s:openapi-gen=target
 type openAPI struct {
-	config       *Config
-	swagger      *spec.Swagger
-	protocolList []string
+	config             *Config
+	swagger            *spec.Swagger
+	protocolList       []string
+	openAPIDefinitions *generators.OpenAPIDefinitions
 }
 
 // RegisterOpenAPIService registers a handler to provides standard OpenAPI specification.
 func RegisterOpenAPIService(config *Config, containers *restful.Container) (err error) {
-	var _ = loads.Spec
-	var _ = strfmt.ParseDuration
-	var _ = validate.FormatOf
 	o := openAPI{
 		config: config,
+		swagger: &spec.Swagger{
+			SwaggerProps: spec.SwaggerProps{
+				Swagger:     OpenAPIVersion,
+				Definitions: spec.Definitions{},
+				Paths:       &spec.Paths{Paths: map[string]spec.PathItem{}},
+				Info:        config.Info,
+			},
+		},
 	}
-	err = o.buildSwaggerSpec()
+
+	err = o.init()
 	if err != nil {
 		return err
 	}
-	containers.ServeMux.HandleFunc(OpenAPIServePath, func(w http.ResponseWriter, r *http.Request) {
+
+	containers.ServeMux.HandleFunc(config.OpenAPIServePath, func(w http.ResponseWriter, r *http.Request) {
 		resp := restful.NewResponse(w)
-		if r.URL.Path != OpenAPIServePath {
+		if r.URL.Path != config.OpenAPIServePath {
 			resp.WriteErrorString(http.StatusNotFound, "Path not found!")
 		}
+		// TODO: we can cache json string and return it here.
 		resp.WriteAsJson(o.swagger)
 	})
 	return nil
 }
 
-func (o *openAPI) buildSwaggerSpec() (err error) {
-	if o.swagger != nil {
-		return fmt.Errorf("Swagger spec is already built. Duplicate call to buildSwaggerSpec is not allowed.")
-	}
-	o.protocolList, err = o.buildProtocolList()
-	if err != nil {
-		return err
-	}
-	definitions, err := o.buildDefinitions()
-	if err != nil {
-		return err
+func (o *openAPI) init() error {
+	if o.openAPIDefinitions == nil {
+		if openAPIGetType, ok := interface{}(o).(interface {
```

Comment:
I guess this is so that you get a run-time error instead of a compile time error if you forget to run the generator? This is worth a comment, probably.